### PR TITLE
[GLUTEN-7145][CH] Decouple `SerializedPlanParser` from other parser modules

### PR DESCRIPTION
**Pull request by lgbo-ustc**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

Fixes: #7145

Decouple `SerializedPlanParser` from all other modules. Introduce `ParserContext` and `ExpressionParser` in this PR. 

 `ParserContext` is the minimal core that need be shared among parser modules. 

`ExpressionParser` is responsilbe for parser substrait expession into `CH` actions dag. It unifies this tranformation among all parser modules. Base on this, some optimization will be applied at later.

 All these should make the code more flexible. 

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

unit tests

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: cpp-ch/local-engine/Parser/ParserContext.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    std::optional<String> getFunctionSignature(Int32 ref) const;
+    std::optional<String> getFunctionSignature(const substrait::Expression_ScalarFunction & func) const;
+    std::optional<String> getFunctionNameInSignature(Int32 ref) const;
+    std::optional<String> getFunctionNameInSignature(const substrait::Expression_ScalarFunction & func) const;
+    const std::unordered_map<String, String> & getLegacyFunctionMapping() const;
+
+private:
+    DB::ContextPtr query_context;
+    std::unordered_map<String, String> legacy_function_mapping;
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Parser/ParserContext.h:55]: use reference may be better

**Review by liuneng1994 — COMMENTED** [review]
LGTM

**Comment by baibaichen** [comment]
@lgbo-ustc please hold this PR

**Comment by lgbo-ustc** [comment]
> @lgbo-ustc please hold this PR

?

**Review by baibaichen — APPROVED** [review]
LGTM

**Pull request merged**